pkg/trustedresources/verify.go

```diff
@@ -100,18 +89,99 @@ func VerifyPipeline(ctx context.Context, pipelineObj v1beta1.PipelineObject, k8s
 		ObjectMeta: pm,
 		Spec:       pipelineObj.PipelineSpec(),
 	}
-	verifiers, err := getVerifiers(ctx, k8s)
+
+	return verifyResource(ctx, &pipeline, k8s, signature, source, policies)
+}
+
+// verifyResource verifies resource which implements metav1.Object by provided signature and public keys from configmap or policies.
+// It will fetch public key from configmap first, if no keys are found then try to fetch keys from VerificationPolicy
+// For verificationPolicies verifyResource will adopt the following rules to do verification:
+// 1. For each policy, check if the resource url is matching any of the `patterns` in the `resources` list. If matched then this policy will be used for verification.
+// 2. If multiple policies are matched, the resource needs to pass all of them to pass verification.
+// 3. To pass one policy, the resource can pass any public keys in the policy.
+func verifyResource(ctx context.Context, resource metav1.Object, k8s kubernetes.Interface, signature []byte, source string, policies []*v1alpha1.VerificationPolicy) error {
+	verifiers := verifiersFromConfigMap(ctx, k8s)
+	if len(verifiers) != 0 {
+		for _, verifier := range verifiers {
+			if err := verifyInterface(resource, verifier, signature); err == nil {
+				return nil
+			}
+		}
+		return fmt.Errorf("resource %s in namespace %s fails verification", resource.GetName(), resource.GetNamespace())
+	}
+
+	if len(policies) == 0 {
+		return errorEmptyVerificationConfig
+	}
+
+	matchedPolicies := matchedPolicies(policies, source)
+	if len(matchedPolicies) == 0 {
+		return fmt.Errorf("no matching policies are found for resource: %s against source: %s. %w", resource.GetName(), source, errorNoMatchedPolicies)
+	}
+
+	verifiedPolicies, err := verifiedPolicies(ctx, resource, k8s, signature, matchedPolicies)
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to verify the resource with matched policies: %w", err)
+	}
+
+	// make sure all matched policies pass verification
+	if verifiedPolicies == len(matchedPolicies) {
```

Comment:
i find it a bit confusing that this is comparing the number of policies to the expected number - i guess because you pass matchedPolicies into the function that counts the number that match this should work, but i wonder if it would be clearer if you verified the matched policies one at a time? i.e. move the `for _, p := range policies {` loop into this function

Reply:
fixed